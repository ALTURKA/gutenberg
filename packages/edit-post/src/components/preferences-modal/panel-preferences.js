/**
 * External dependencies
 */
import { get, filter, map } from 'lodash';

/**
 * WordPress dependencies
 */
import { store as coreStore } from '@wordpress/core-data';
import { useSelect } from '@wordpress/data';
import {
	PostTaxonomies,
	PostExcerptCheck,
	PageAttributesCheck,
	PostFeaturedImageCheck,
	PostTypeSupportCheck,
	store as editorStore,
} from '@wordpress/editor';
import { __ } from '@wordpress/i18n';
import { PreferencesModalSection } from '@wordpress/interface';

/**
 * Internal dependencies
 */
import {
	EnableCustomFieldsOption,
	EnablePluginDocumentSettingPanelOption,
	EnablePanelOption,
} from './options';
import { store as editPostStore } from '../../store';

export function MetaBoxesSection() {
	const { areCustomFieldsRegistered, metaBoxes } = useSelect( ( select ) => {
		const { getEditorSettings } = select( editorStore );
		const { getAllMetaBoxes } = select( editPostStore );

		return {
			// This setting should not live in the block editor's store.
			areCustomFieldsRegistered:
				getEditorSettings().enableCustomFields !== undefined,
			metaBoxes: getAllMetaBoxes(),
		};
	} );
	// The 'Custom Fields' meta box is a special case that we handle separately.
	const thirdPartyMetaBoxes = filter(
		metaBoxes,
		( { id } ) => id !== 'postcustom'
	);

	if ( ! areCustomFieldsRegistered && thirdPartyMetaBoxes.length === 0 ) {
		return null;
	}

	return (
		<PreferencesModalSection
			title={ __( 'Additional' ) }
			description={ __( 'Add extra areas to the editor.' ) }
		>
			{ areCustomFieldsRegistered && (
				<EnableCustomFieldsOption label={ __( 'Custom fields' ) } />
			) }
			{ map( thirdPartyMetaBoxes, ( { id, title } ) => (
				<EnablePanelOption
					key={ id }
					label={ title }
					panelName={ `meta-box-${ id }` }
				/>
			) ) }
		</PreferencesModalSection>
	);
}

export default function PanelPreferences() {
	const isViewable = useSelect( ( select ) => {
		const { getEditedPostAttribute } = select( editorStore );
		const { getPostType } = select( coreStore );
		const postType = getPostType( getEditedPostAttribute( 'type' ) );
		return get( postType, [ 'viewable' ], false );
	}, [] );

	return (
		<>
			<PreferencesModalSection
				title={ __( 'Document settings' ) }
				description={ __( 'Choose what displays in the panel.' ) }
			>
				<EnablePluginDocumentSettingPanelOption.Slot />
				{ isViewable && (
					<EnablePanelOption
						label={ __( 'Permalink' ) }
						panelName="post-link"
					/>
				) }
				{ isViewable && (
					<EnablePanelOption
						label={ __( 'Template' ) }
						panelName="template"
					/>
				) }
				<PostTaxonomies
					taxonomyWrapper={ ( content, taxonomy ) => (
						<EnablePanelOption
							label={ get( taxonomy, [ 'labels', 'menu_name' ] ) }
							panelName={ `taxonomy-panel-${ taxonomy.slug }` }
						/>
					) }
				/>
				<PostFeaturedImageCheck>
					<EnablePanelOption
						label={ __( 'Featured image' ) }
						panelName="featured-image"
					/>
				</PostFeaturedImageCheck>
				<PostExcerptCheck>
					<EnablePanelOption
						label={ __( 'Excerpt' ) }
						panelName="post-excerpt"
					/>
				</PostExcerptCheck>
				<PostTypeSupportCheck
					supportKeys={ [ 'comments', 'trackbacks' ] }
				>
					<EnablePanelOption
						label={ __( 'Discussion' ) }
						panelName="discussion-panel"
					/>
				</PostTypeSupportCheck>
				<PageAttributesCheck>
					<EnablePanelOption
						label={ __( 'Page attributes' ) }
						panelName="page-attributes"
					/>
				</PageAttributesCheck>
			</PreferencesModalSection>
			<MetaBoxesSection />
		</>
	);
}
