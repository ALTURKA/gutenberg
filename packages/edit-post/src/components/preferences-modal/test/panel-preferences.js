/**
 * External dependencies
 */
import { shallow } from 'enzyme';

/**
 * WordPress dependencies
 */
import { useSelect } from '@wordpress/data';

/**
 * Internal dependencies
 */
import { MetaBoxesSection } from '../panel-preferences';

jest.mock( '@wordpress/data/src/components/use-select', () => jest.fn() );

describe( 'MetaBoxesSection', () => {
	it( 'does not render if there are no options', () => {
		useSelect.mockImplementation( () => ( {
			areCustomFieldsRegistered: false,
			metaBoxes: [
				{ id: 'postcustom', title: 'This should not render' },
			],
		} ) );

		const wrapper = shallow( <MetaBoxesSection /> );
		expect( wrapper.isEmptyRender() ).toBe( true );
	} );

	it( 'renders a Custom Fields option', () => {
		useSelect.mockImplementation( () => ( {
			areCustomFieldsRegistered: true,
			metaBoxes: [
				{ id: 'postcustom', title: 'This should not render' },
			],
		} ) );

		const wrapper = shallow( <MetaBoxesSection /> );
		expect( wrapper ).toMatchSnapshot();
	} );

	it( 'renders meta box options', () => {
		useSelect.mockImplementation( () => ( {
			areCustomFieldsRegistered: false,
			metaBoxes: [
				{ id: 'postcustom', title: 'This should not render' },
				{ id: 'test1', title: 'Meta Box 1' },
				{ id: 'test2', title: 'Meta Box 2' },
			],
		} ) );

		const wrapper = shallow( <MetaBoxesSection /> );
		expect( wrapper ).toMatchSnapshot();
	} );

	it( 'renders a Custom Fields option and meta box options', () => {
		useSelect.mockImplementation( () => ( {
			areCustomFieldsRegistered: true,
			metaBoxes: [
				{ id: 'postcustom', title: 'This should not render' },
				{ id: 'test1', title: 'Meta Box 1' },
				{ id: 'test2', title: 'Meta Box 2' },
			],
		} ) );

		const wrapper = shallow( <MetaBoxesSection /> );
		expect( wrapper ).toMatchSnapshot();
	} );
} );
