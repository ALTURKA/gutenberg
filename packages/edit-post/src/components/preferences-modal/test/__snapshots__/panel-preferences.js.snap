// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetaBoxesSection renders a Custom Fields option 1`] = `
<PreferencesModalSection
  description="Add extra areas to the editor."
  title="Additional"
>
  <WithSelect(EnableCustomFieldsOption)
    label="Custom fields"
  />
</PreferencesModalSection>
`;

exports[`MetaBoxesSection renders a Custom Fields option and meta box options 1`] = `
<PreferencesModalSection
  description="Add extra areas to the editor."
  title="Additional"
>
  <WithSelect(EnableCustomFieldsOption)
    label="Custom fields"
  />
  <WithSelect(IfCondition(WithDispatch(PreferencesModalToggle)))
    key="test1"
    label="Meta Box 1"
    panelName="meta-box-test1"
  />
  <WithSelect(IfCondition(WithDispatch(PreferencesModalToggle)))
    key="test2"
    label="Meta Box 2"
    panelName="meta-box-test2"
  />
</PreferencesModalSection>
`;

exports[`MetaBoxesSection renders meta box options 1`] = `
<PreferencesModalSection
  description="Add extra areas to the editor."
  title="Additional"
>
  <WithSelect(IfCondition(WithDispatch(PreferencesModalToggle)))
    key="test1"
    label="Meta Box 1"
    panelName="meta-box-test1"
  />
  <WithSelect(IfCondition(WithDispatch(PreferencesModalToggle)))
    key="test2"
    label="Meta Box 2"
    panelName="meta-box-test2"
  />
</PreferencesModalSection>
`;
